### DISCUSSION TITLE
Issue with fprime-util build aarch64-linux

### DISCUSSION THREAD

**Discussion by OwenNieuwenhuizen** [nasa/fprime#2682] [Q&A]
Hi,

I am trying to build for a raspberry pi and I am using a library that sends pwm. I am getting this error when I try to build with fprime-util build aarch64-linux.

[100%] Linking CXX executable ../bin/aarch64-linux/GDep
Toolchain 'aarch64-linux' should set CMAKE_SYSTEM_NAME to 'Generic' and set
FPRIME_PLATFORM
Call Stack (most recent call first):
fprime/cmake/FPrime.cmake:29 (include)
CMakeLists.txt:13 (include)

/opt/toolchains/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /usr/lib/libwiringPi.so: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
gmake[3]: *** [GDep/CMakeFiles/GDep.dir/build.make:217: bin/aarch64-linux/GDep] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:5288: GDep/CMakeFiles/GDep.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2
gmake: *** [Makefile:1339: GDep] Error 2
[ERROR] CMake erred with return code 2
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep#

It gives me this warning at the top:

(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep# fprime-util build aarch64-linux
-- [python3] python3 found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/python3
-- [fpp-tools] fpp-depend found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-depend
-- [fpp-tools] fpp-to-xml found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-xml
-- [fpp-tools] fpp-to-cpp found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-cpp
-- [fpp-tools] fpp-locate-defs found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-locate-defs
-- [FPRIME] Module locations: /mnt/c/Users/owenn/Command/fprime;/mnt/c/Users/owenn/Command
Linux/mnt/c/Users/owenn/Command/fprime/cmake/toolchain/aarch64-linux.cmake
aarch64-linux
CMake Warning at fprime/cmake/platform/platform.cmake:28 (message):
-- [FPRIME] Configuration module: /mnt/c/Users/owenn/Command/fprime/config
-- [FPRIME] Installation directory: /mnt/c/Users/owenn/Command/build-artifacts
-- Target build toolchain/platform: aarch64-linux/Linux
-- Including /mnt/c/Users/owenn/Command/fprime/cmake/platform/Linux.cmake
-- Requiring thread library

Any ideas? I am thinking it could possibly be a linker error or a variable not being set properly, but I am new to this so I am unsure. I have a mission using this software coming up quickly though so after spending quite a bit of time so far on trying to debug this, I would greatly appreciate any advice.

**Comment by tide999** [comment]
----- 原始邮件 -----
发件人：OwenNieuwenhuizen ***@***.***>
收件人：nasa/fprime ***@***.***>
抄送人：Subscribed ***@***.***>
主题：[nasa/fprime] Issue with fprime-util build aarch64-linux (Discussion #2682)
日期：2024年04月14日 09点34分


Hi,
I am trying to build for a raspberry pi and I am using a library that sends pwm. I am getting this error when I try to build with fprime-util build aarch64-linux.
[100%] Linking CXX executable ../bin/aarch64-linux/GDep

Toolchain 'aarch64-linux' should set CMAKE_SYSTEM_NAME to 'Generic' and set

FPRIME_PLATFORM

Call Stack (most recent call first):

fprime/cmake/FPrime.cmake:29 (include)

CMakeLists.txt:13 (include)
/opt/toolchains/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /usr/lib/libwiringPi.so: error adding symbols: file in wrong format

collect2: error: ld returned 1 exit status

gmake[3]: *** [GDep/CMakeFiles/GDep.dir/build.make:217: bin/aarch64-linux/GDep] Error 1

gmake[2]: *** [CMakeFiles/Makefile2:5288: GDep/CMakeFiles/GDep.dir/all] Error 2

gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2

gmake: *** [Makefile:1339: GDep] Error 2

[ERROR] CMake erred with return code 2

(fprime-venv) ***@***.***:/mnt/c/Users/owenn/Command/GDep#
It gives me this warning at the top:
(fprime-venv) ***@***.***:/mnt/c/Users/owenn/Command/GDep# fprime-util build aarch64-linux

-- [python3] python3 found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/python3

-- [fpp-tools] fpp-depend found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-depend

-- [fpp-tools] fpp-to-xml found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-xml

-- [fpp-tools] fpp-to-cpp found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-to-cpp

-- [fpp-tools] fpp-locate-defs found at: /mnt/c/Users/owenn/Command/fprime-venv/bin/fpp-locate-defs

-- [FPRIME] Module locations: /mnt/c/Users/owenn/Command/fprime;/mnt/c/Users/owenn/Command

Linux/mnt/c/Users/owenn/Command/fprime/cmake/toolchain/aarch64-linux.cmake

aarch64-linux

CMake Warning at fprime/cmake/platform/platform.cmake:28 (message):

-- [FPRIME] Configuration module: /mnt/c/Users/owenn/Command/fprime/config

-- [FPRIME] Installation directory: /mnt/c/Users/owenn/Command/build-artifacts

-- Target build toolchain/platform: aarch64-linux/Linux

-- Including /mnt/c/Users/owenn/Command/fprime/cmake/platform/Linux.cmake

-- Requiring thread library
Any ideas? I am thinking it could possibly be a linker error or a variable not being set properly, but I am new to this so I am unsure. I have a mission using this software coming up quickly though so after spending quite a bit of time so far on trying to debug this, I would greatly appreciate any advice.

**Reply by OwenNieuwenhuizen** [reply]
?

**Reply by OwenNieuwenhuizen** [reply]
Update: I looked into it a bit more and it doesn't look like a linking error, but more likely a compatibility error. I did some research on the library I am using, WiringPi, and it looks like it is built for a 32-bit architecture instead of a 64-bit, meaning that it will not allow the deployment to build given that it is not meant for that architecture. I found a set of instructions on how to upgrade WiringPi to a 64-bit architecture, but I haven't had much luck with that so far. This is the website with the instructions in case anyone is curious or has any advice on the subject: https://roboticsbackend.com/introduction-to-wiringpi-for-raspberry-pi/.

**Comment by LeStarch** [comment]
Yes, the error you see is an incompatible version of wiringPI.  You can try to upgrade wiringPI .... Or the raspberry PI can often run 32bit binaries.

You might try compiling with 32bit cross compiler instead of upgrading the library.

**Reply by OwenNieuwenhuizen** [reply]
I went a different route and redid it using pigpio instead. It successfully built on the default, and it generated for aarch64-linux, but when I tried to build for aarch64-linux, it gave me this error:

/mnt/c/Users/owenn/Command/Components/Gimbal/Gimbal.cpp:9:10: fatal error: pigpio.h: No such file or directory
    9 | #include <pigpio.h> // Include PiGpio Library
      |          ^~~~~~~~~~
compilation terminated.
gmake[3]: *** [Components/Gimbal/CMakeFiles/Components_Gimbal.dir/build.make:105: Components/Gimbal/CMakeFiles/Components_Gimbal.dir/Gimbal.cpp.o] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:5177: Components/Gimbal/CMakeFiles/Components_Gimbal.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2
gmake: *** [Makefile:1339: GDep] Error 2
[ERROR] CMake erred with return code 2

I think it for some reason is not recognizing the h file for the pigpio functions. What would you recommend I do to fix this. Is it some kind of linking error. I currently have the pigpio library as a first layer directory in my project (at same level as my deployment). What would you recommend?

**Reply by OwenNieuwenhuizen** [reply]
Actually I fixed that error. I forgot to include_directories. Now I am running into a different error at the very end of the build process unfortunately. I think that it is saying that pigpio library is also somehow in the wrong format. I'm guessing it is saying that it is in an x32 format instead of an x64. How would you recommend I fix this?

[100%] Linking CXX executable ../bin/aarch64-linux/GDep
CMake Warning at fprime/cmake/platform/platform.cmake:28 (message):
  Toolchain 'aarch64-linux' should set CMAKE_SYSTEM_NAME to 'Generic' and set
  FPRIME_PLATFORM
Call Stack (most recent call first):
  fprime/cmake/FPrime.cmake:29 (include)
  CMakeLists.txt:13 (include)


/opt/toolchains/bin/../lib/gcc/aarch64-none-linux-gnu/10.2.1/../../../../aarch64-none-linux-gnu/bin/ld: /mnt/c/Users/owenn/Command/pigpio/libpigpio.so: error adding symbols: file in wrong format
collect2: error: ld returned 1 exit status
gmake[3]: *** [GDep/CMakeFiles/GDep.dir/build.make:217: bin/aarch64-linux/GDep] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:5288: GDep/CMakeFiles/GDep.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:5295: GDep/CMakeFiles/GDep.dir/rule] Error 2
gmake: *** [Makefile:1339: GDep] Error 2
[ERROR] CMake erred with return code 2
(fprime-venv) root@OwenPC:/mnt/c/Users/owenn/Command/GDep#
